providers/fips/fipsprov.c

```diff
@@ -37,17 +37,19 @@ static OSSL_FUNC_provider_get_params_fn fips_get_params;
 static OSSL_FUNC_provider_query_operation_fn fips_query;
 
 /* Locale object accessor functions */
-#ifdef OPENSSL_SYS_MACOSX
+#if defined (OPENSSL_SYS_MACOSX)
 # include <xlocale.h>
 #else
 # include <locale.h>
 #endif
 
-#if defined OPENSSL_SYS_WINDOWS
+#if defined (OPENSSL_SYS_WINDOWS)
 # define locale_t _locale_t
 # define freelocale _free_locale
 #endif
+#if !defined (OPENSSL_SYS_TANDEM)
```

Comment:
This should use the condition as in ctype.c and I'd also add a define to be used later:
```suggestion
#if ((defined OPENSSL_SYS_WINDOWS) || (defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200809L)) && !defined (OPENSSL_SYS_TANDEM)
# define OPENSSL_USE_LOCALE
```

Reply:
@t8m With the report from @florinspatar:
> Same issue on HP-UX, AIX, Solaris and older Linux (RHEL5 & SLES10). Also tested with the proposed patch, but there is no difference.

this has become a much broader issue and is not specific to NonStop. I agree with the suggestion to use `OPENSSL_USE_LOCALE` instead of what @beldmit proposed because `#if` at line 50 is insufficient given the new scope of the breakage. I am tempted to withdraw the PR until a more general solution is put in place, which likely belongs in the Configuration infrastructure rather than detected in the above line 50. So a platform should define be something like `OPENSSL_SUPPRESS_LOCALE` if locale is unavailable. For example, `Configurations/50-nonstop.pm` would contain:

```
        defines          => add('OPENSSL_VPROC=$(OPENSSL_VPROC)',
                                '_XOPEN_SOURCE',
                                '_XOPEN_SOURCE_EXTENDED=1',
                                '_TANDEM_SOURCE',
                                'B_ENDIAN',
                                'OPENSSL_SUPPRESS_LOCALE`),
```

Reply:
`_XOPEN_SOURCE` with no value exposes features the NonStop libraries don't seem to support.
From reading the [feature_test_macros(7)](https://man7.org/linux/man-pages/man7/feature_test_macros.7.html), `_XOPEN_SOURCE=600` should be the right level (supports POSIX.1-2001, but not POSIX.1-2008, the latter being where `locale_t` etc are defined)

Reply:
I checked the header files many times through this issue. No `locale_t` is present on NonStop at any version for C. There is a `locale` include for C++ but it is not usable for C programs. The new/free routines are definitely not present in any form. The presence of locale_t appears to be an optional POSIX component. Regardless, this is a new dependency in the 3.0.x series and I am trying to figure out some way around this problem.

Reply:
> @t8m With the report from @florinspatar:
> 
> > Same issue on HP-UX, AIX, Solaris and older Linux (RHEL5 & SLES10). Also tested with the proposed patch, but there is no difference.

Are we sure this isn't just the problem with the fipsprov.c not being right at all in regards to the posix version check?

Reply:
Could you please apply the suggested changes, then we can ask @florinspatar to try this patch.

Reply:
I am working on them. It takes time to run a build/test. Please stay tuned.

Reply:
Unfortunately the changes as done will not solve the missing POSIX test in fipsprov.c.

Reply:
> Unfortunately the changes as done will not solve the missing POSIX test in fipsprov.c.

Would you mind clarifying what is needed? This does fix the breakage, which was my the priority. I think @beldmit should chime in on it since he suggested the original fix in fipsprov.c, which I made work.

Reply:
Yes. My intention was to help fix your issue. The correct check should probably rely on OPENSSL_SUPPRESS_LOCALE if you suggest it as more universal define

Reply:
> Yes. My intention was to help fix _your_ issue. The correct check should probably rely on OPENSSL_SUPPRESS_LOCALE if you suggest it as more universal define

That's what I pushed a moment ago. I called it OPENSSL_SUPPRESS_LOCALE_T to be more specific. The approach does not appear to be acceptable, however.